src/test/java/org/broadinstitute/hellbender/engine/dataflow/transforms/RefBasesFromAPIUnitTest.java

```diff
@@ -110,4 +117,56 @@ public void noReadsTest(List<KV<ReferenceShard, Iterable<GATKRead>>> kvRefShardi
 
         p.run();
     }
+
+    @DataProvider(name = "RefBasesFromAPIWithCustomWindowFunctionTestData")
+    public Object[][] refBasesFromAPIWithCustomWindowFunctionTestData() {
+        final ReferenceShard shard = new ReferenceShard(0, "1");
+        final List<GATKRead> samReads = Arrays.asList(
+                ReadsPreprocessingPipelineTestData.makeRead("1", 500, 100, 0, SAMRecord.class),
+                ReadsPreprocessingPipelineTestData.makeRead("1", 1000, 100, 1, SAMRecord.class),
+                ReadsPreprocessingPipelineTestData.makeRead("1", 1050, 100, 2, SAMRecord.class)
+        );
+        final List<GATKRead> googleReads = Arrays.asList(
+                ReadsPreprocessingPipelineTestData.makeRead("1", 500, 100, 0, Read.class),
+                ReadsPreprocessingPipelineTestData.makeRead("1", 1000, 100, 1, Read.class),
+                ReadsPreprocessingPipelineTestData.makeRead("1", 1050, 100, 2, Read.class)
+        );
+
+        final List<Object[]> testCases = new ArrayList<>();
+        for ( final List<GATKRead> reads : Arrays.asList(samReads, googleReads) ) {
+            // Test case layout: input shard + reads, reference window function to apply, expected reference interval post-transform
+            testCases.add( new Object[]{ KV.of(shard, reads), RefWindowFunctions.IDENTITY_FUNCTION, new SimpleInterval("1", 500, 1149) });
+            testCases.add( new Object[]{ KV.of(shard, reads), new RefWindowFunctions.FixedWindowFunction(1, 1), new SimpleInterval("1", 499, 1150) });
+            testCases.add( new Object[]{ KV.of(shard, reads), new RefWindowFunctions.FixedWindowFunction(3, 5), new SimpleInterval("1", 497, 1154) });
+            testCases.add( new Object[]{ KV.of(shard, reads), new RefWindowFunctions.FixedWindowFunction(30, 10), new SimpleInterval("1", 470, 1159) });
+            testCases.add( new Object[]{ KV.of(shard, reads), new RefWindowFunctions.FixedWindowFunction(1000, 1000), new SimpleInterval("1", 1, 2149) });
+        }
+        return testCases.toArray(new Object[][]{});
+    }
+
+    @Test(dataProvider = "RefBasesFromAPIWithCustomWindowFunctionTestData")
+    public void testRefBasesFromAPIWithCustomWindowFunction( final KV<ReferenceShard, Iterable<GATKRead>> inputShard, final SerializableFunction<GATKRead, SimpleInterval> referenceWindowFunction, final SimpleInterval expectedReferenceInterval ) {
+        Pipeline p = GATKTestPipeline.create();
+        DataflowUtils.registerGATKCoders(p);
+
+        // Assuming everything works properly, we only need a mock response to the expected reference interval,
+        // since that is what should end up being queried in RefBasesFromAPI. If we later try to query an
+        // unexpected interval, we'll detect it via an Assert.
```

Comment:
I've verified that the assert actually does catch this case (mock returning null), at least for this test. For the higher-level transforms like `AddContextDataToRead` you're right, though -- transform blows up before you get a chance to assert. This is why I wish we could use `Answer` with mockito...